<p><b>LDlink 5.6.3 Release (6/1/2023)</b></p>
<ul>
    <li>Fixed LDassoc Upload Issues</li>
</ul>
<br>
<p><b>LDlink 5.6.2 Release (5/15/2023)</b></p>
<ul>
    <li>Domain URL migrated from ldlink.nci.nih.gov to ldlink.nih.gov</li>
    <li>Application logo changed</li>
</ul>
<br>
<p><b>LDlink 5.6.1 Release (5/3/2023)</b></p>
<ul>
    <li>Fixed download button which can not load file in new tab for Chrome and Firefox</li>
</ul>
<br>
<p><b>LDlink 5.6.0 Release (4/20/2023)</b></p>
<ul>
    <li>Migrated to serverless architecture for improved performance, reliability, and scalability.</li>
</ul>
<br>
<p><b>LDlink 5.5.2 Release (3/25/2023)</b></p>
<ul>
    <li>Fix the multiallelic error message for LDpair</li>
</ul>
<br>
<p><b>LDlink 5.5.1 Release (11/15/2022)</b></p>
<ul>
    <li>Add a link to FORGEdb scoring scheme for LDassoc and LDproxy</li>
</ul>
<br>
<p><b>LDlink 5.5.0 Release (11/03/2022)</b></p>
<ul>
    <li>Publicly available scores from FORGEdb are used by LDlink's LDassoc and LDproxy and LDMatrix
    modules. </li>
    <li>Added Global Diversity Array_Confluence (I_GDA-C) and Illumina Global Screening version 3_Confluence (I_GSA-v3C) to
    SNPChip.</li>
    <li>Provide a warning when the input on LDtrait can potentially cause a server timeout.</li> 
    <li>Fixed the broken link for View R2 data in UCGC Genome Browse</li>
</ul>
<br>
<p><b>LDlink 5.4.2 Release (09/13/2022)</b></p>
<ul>
    <li>Fix the output format for LDpair. It should be text file for api call</li>
</ul>
<br>
<p><b>LDlink 5.4.1 Release (09/01/2022)</b></p>
<ul>
    <li>Provide more useful error message for LDpair if varient is not in 1000G</li>
</ul>
<br>
<p><b>LDlink 5.4.0 Release (08/08/2022)</b></p>
<ul>
    <li>Internal code refactoring for performance improvements</li>
</ul>
<br>
<p><b>LDlink 5.3.3 Release (05/24/2022)</b></p>
<ul>
    <li>Revise SNPclip sort order, use user supplied input format;</li> 
    <li>Fix LDpair multiallelic query for second SNP input</li>
</ul>
<br>
<p><b>LDlink 5.3.2 Release (05/06/2022)</b></p>
<ul>
    <li>Fix LDmatrix plotting for unsorted query variants.</li>
</ul>
<br>
<p><b>LDlink 5.3.1 Release (04/28/2022)</b></p>
<ul>
    <li>Improve ability to catch SNPs missing in 1000G data.</li>
</ul>
<br>
<p><b>LDlink 5.3 Release (04/05/2022)</b></p>
<ul>
    <li>Add ability to submit batch LDpair inputs to single API requests.</li>
    <li>Fix behavior when a SNP's 1000G position does not match dbSNP.</li>
</ul>
<br>
<p><b>LDlink 5.2 Release (01/03/2022)</b></p>
<ul>
    <li>dbSNP build updated from version 151 to 155.</li>
    <li>Add support for GRCh38 genomic coordinates.</li>
    <li>Add support for genetic map recombination rates in chromosome X.</li>
    <li>Add ability to collapsed gene transcripts in LDmatrix and LDproxy.</li>
</ul>
<br>
<p><b>LDlink 5.1 Release (05/25/2021)</b></p>
<ul>
    <li>Improved filtering options on LDexpress data table.</li>
    <li>Allow scientific notation as input for LDexpress P-value threshold.</li>
</ul>
<br>
<p><b>LDlink 5.0 Release (11/06/2020)</b></p>
<ul>
    <li><b>LDexpress</b> released for searching variants that are associated with gene expression in multiple tissue types.</li>
</ul>
<br>
<p><b>LDlink 4.2.0 Release (09/21/2020)</b></p>
<ul>
    <li>Add ability to LDproxy to adjust base pair window size.</li>
</ul>
<br>
<p><b>LDlink 4.1.0 Release (04/29/2020)</b></p>
<ul>
    <li>Add ability to LDtrait to adjust base pair window size.</li>
</ul>
<br>
<p><b>LDlink 4.0.3 Release (04/06/2020)</b></p>
<ul>
    <li>Fix population dropdown UI cut short by module container.</li>
</ul>
<br>
<p><b>LDlink 4.0.2 Release (03/20/2020)</b></p>
<ul>
    <li>Fix P-value and 95% CI columns in LDtrait to reflect GWAS Catalog.</li>
    <li>Add timestamp of when LDtrait GWAS Catalog was last updated.</li>
    <li>Fix browser forward/back functionality.</li>
</ul>
<br>
<p><b>LDlink 4.0.1 Release (03/10/2020)</b></p>
<ul>
    <li>Fix SNPchip array labels not displaying correctly.</li>
</ul>
<br>
<p><b>LDlink 4.0 Release (03/10/2020)</b></p>
<ul>
    <li><b>LDtrait</b> released for searching variants that have been associated with a trait or disease.</li>
    <li>Fix correlated allele calculation for LDassoc, LDmatrix, LDpair, and LDproxy.</li>
    <li>Combine R<sup>2</sup> and D' coloring for LDmatrix plots.</li>
    <li>Home page redesign.</li>
</ul>
<br>
<p><b>LDlink 3.9 Release (12/10/2019)</b></p>
<ul>
    <li>Add population details, P-value, and Chi-sq to LDpop export table.</li>
</ul>
<br>
<p><b>LDlink 3.8 Release (10/31/2019)</b></p>
<ul>
    <li>Upgrade backend from Python 2 to Python 3.</li>
</ul>
<br>
<p><b>LDlink 3.7.2 Release (08/23/2019)</b></p>
<ul>
    <li>Fix variant inputs by genomic coordinates for chromosomes X and Y.</li>
</ul>
<br>
<p><b>LDlink 3.7.1 Release (07/30/2019)</b></p>
<ul>
    <li>Fix collapsed gene transcripts in LDassoc.</li>
</ul>
<br>
<p><b>LDlink 3.7 Release (05/09/2019)</b></p>
<ul>
    <li>Return LDpop API results in tab-delimited format.</li>
</ul>
<br>
<p><b>LDlink 3.6 Release (04/03/2019)</b></p>
<ul>
    <li>Add ability to make POST requests to LDmatrix with up to 1000 variants.</li>
    <li>Restrict concurrent requests to the API.</li>
</ul>
<br>
<p><b>LDlink 3.5 Release (03/18/2019)</b></p>
<ul>
    <li><b>LDpop</b> released for visualizing LD patterns across 1000G populations.</li>
</ul>
<br>
<p><b>LDlink 3.4 Release (02/06/2019)</b></p>
<ul>
    <li>Correct dbsnp positions for delins, ins, and del type variants.</li>
    <li>Improve query performance from genomic coordinate and rsid inputs.</li>
</ul>
<br>
<p><b>LDlink 3.3 Release (12/24/2018)</b></p>
<ul>
    <li>LDlink API access now requires registration. Please register using the <u>API Access</u> tab</li>
    <li>dbSNP build updated from version 142 to 151.</li>
    <li>Genomic coordinates can now be used to query variants for each module.</li>
    <li>Improvements have been made to error messages from programmatic access.</li>
    <li>Refresh user interface.</li>
</ul>
<br>
<p><b>LDlink 3.2 Release (08/01/2018)</b></p>
<ul>
    <li>Update Bokeh interactive plotting library to release 0.13.0.</li>
    <li>Add ability to export Bokeh plots to high quality images (SVG, PDF, PNG, and JPEG).</li>
    <li>Merge gene plot and main plot into one image during export.</li>
    <li>Fix errors with SNPclip output when variants are not found in 1000G VCF file.</li>
</ul>
<br>
<p><b>LDlink 3.0 Release (04/01/2017)</b></p>
<ul>
    <li><b>LDassoc</b> released for visualizing association results.</li>
    <li><b>Programmatic Access</b> of LDlink via terminal commands.</li>
    <li>Update Bokeh interactive plotting library to release 0.12.2.</li>
    <li>Improve performance of allele matching algorithm.</li>
    <li>Better table sorting on absolute value of distance.</li>
    <li>Refine exported UCSC data tracks (link at bottom of interactive plots).</li>
</ul>
<br>
<p><b>LDlink 2.0 Release (03/10/2016)</b></p>
<ul>
    <li><b>SNPclip</b> released for LD pruning a list of variants.</li>
    <li><b>SNPchip</b> released for finding variants on commercial genotyping arrays.</li>
    <li>Minor tweaks to improve visualization of indel output.</li>
    <li>Correct p-value rounding to zero in <b>LDpair</b>.</li>
    <li>Fix errors when variants map to the same position in 1000G VCF file</li>
</ul>
<br>
<p><b>LDlink 1.1 Release (09/30/2015)</b></p>
<ul>
    <li>Indels are now accepted as input in all <b>LDlink</b> modules.</li>
    <li><b>LDproxy</b> output table now returns all data with options to sort by columns and search output.</li>
    <li><b>LDhap</b> haplotypes are now color coded by alleles for easier viewing.</li>
    <li><b>LDmatrix</b> now includes a R<sup>2</sup>/D' toggle for plotting R<sup>2</sup> or D' matrices.</li>
    <li>Minor tweaks to improve visualization on a variety of device screen sizes.</li>
</ul>